refactor: warn about release pypi deprecation and encourage to use trusted publisher (#707)
ansys/actions: doc/source/release-actions/index.rst

Below, it is assumed that a job `build-library` already created library sources
and wheel artifacts using the environment variable `PACKAGE_NAME`. Also, the job
is defined to not upload files if one already exists thanks to the
`skip-existing` input.

When used to test the release process, you can update the `repository-url` to
"https://test.pypi.org/legacy/".

```
    release-pypi:
    name: "Release to PyPI with trusted publisher approach"
    runs-on: ubuntu-latest
    needs: [build-library]
    # Specifying a GitHub environment is optional, but strongly encouraged
    environment: release
    permissions:
        # IMPORTANT: this permission is mandatory for trusted publishing
        id-token: write
    if: github.event_name == 'push' && contains(github.ref, 'refs/tags')
    steps:
        - name: "Download the library artifacts from build-library step"
        uses: actions/download-artifact@v4
        with:
            name: ${{ env.PACKAGE_NAME }}-artifacts
            path: ${{ env.PACKAGE_NAME }}-artifacts

        - name: "Display the structure of downloaded files"
        shell: bash
        run: ls -R

        - name: "Upload artifacts to PyPI using Trusted Publisher"

>         uses: pypa/gh-action-pypi-publish@v1.10.1

Reviewer: Suggesting to add here the latest version of the action:

```suggestion
        uses: pypa/gh-action-pypi-publish@v1.12.4
```

Author: Nice catch !

.. jinja:: release-github
    :file: _templates/action.rst.jinja

Release PyPI template
---------------------

Below, it is assumed that a job `build-library` already created library sources
and wheel artifacts using the environment variable `PACKAGE_NAME`. Also, the job
is defined to not upload files if one already exists thanks to the

> `skip-existing` input.

Reviewer 2: I think we need to explain a bit more why we need to change to having it separated in the workflow... currently this doesn't explain the reasons, right?

Author: You're right, a quick explanation is triggered in the pypi-release actions but it could be added here and extended.

Reviewer: I moved this to the migration guide. We have a dedicated page for it. It was created when introducing the trusted publishers for the first time.

Also, I added warnings in the description of every release action. Now, those are included in the docs too.

> Below, it is assumed that a job `build-library` already created library sources
> and wheel artifacts using the environment variable `PACKAGE_NAME`. Also, the job
> is defined to not upload files if one already exists thanks to the
> `skip-existing` input.

Reviewer 3: ```suggestion
The following code assumes that a `build-library` job has already created library sources
and wheel artifacts using the `PACKAGE_NAME` environment variable. Also, the job's 
`skip-existing` input indicates that files that already exist are not uploaded again.
```
It's clearer when the noun describing the code entity follows the code entity. Also, we should avoid placement indicators like above and below and use preceding and following instead. This is because a PDF of the documentation might have placed content on a preceding or following page.

.. code::yaml

    release-pypi:
    name: "Release to PyPI with trusted publisher approach"
    runs-on: ubuntu-latest
    needs: [build-library]

>     # Specifying a GitHub environment is optional, but strongly encouraged

Reviewer 3: ```suggestion
    # Specifying a GitHub environment is optional but strongly encouraged
```

    environment: release
    permissions:
        # IMPORTANT: this permission is mandatory for trusted publishing
        id-token: write
    if: github.event_name == 'push' && contains(github.ref, 'refs/tags')
    steps:

>         - name: "Download the library artifacts from build-library step"

Reviewer 2: ```suggestion
      - name: "Download the library artifacts from build-library step"
```

        uses: actions/download-artifact@v4
        with:
            name: ${{ env.PACKAGE_NAME }}-artifacts
            path: ${{ env.PACKAGE_NAME }}-artifacts

>         - name: "Display the structure of downloaded files"

Reviewer 2: ```suggestion
      - name: "Display the structure of downloaded files"
```

        shell: bash
        run: ls -R

>         - name: "Upload artifacts to PyPI using Trusted Publisher"

Reviewer 2: ```suggestion
      - name: "Upload artifacts to PyPI using Trusted Publisher"
```

Reviewer 3: Replace "Trusted Publisher" with "trusted pulisher approach"?

Reviewer: Just added your feedback to the suggestion.

        uses: pypa/gh-action-pypi-publish@v1.12.4
        with:
            repository-url: "https://upload.pypi.org/legacy/"
            print-hash: true
            packages-dir: ${{ env.PACKAGE_NAME }}-artifacts
            skip-existing: false

> When used to test the release process, you can update the repository URL as
> follow:

Reviewer 3: ```suggestion
When testing the release process, you can update the repository URL as
follows:
```
I hope I didn't change the meaning, but the subject "you" isn't related to "When used to test the release process, ...." For the existing construction, you would need to have something like this: "When used to test the release process, the following action updates the repository URL."
